File: lab5.rb
K = 4127;

class RSA
    @@x, @@y = 0, 0

    def initialize()
        @p,@q = generateRandomNumbers
        @n = @p*@q
        @phi = (@p-1)*(@q-1)
        @e = 31
        @d = inverse_element(@e, @phi)
     end

    def encryption(message)
        message ** @e % @n
    end

    def decription(ciphertext)
        ciphertext ** @d % @n
    end

    def miller_rabin_prime?(n, g)
        d = n - 1
        s = 0
        a = 0
        while d % 2 == 0
            d /= 2
            s += 1
        end
        g.times do |iterator|
            a = 2 + rand(n - 4)
            x = (a**d) % n
            next if x == 1 || x == n - 1
            for r in (1..s - 1)
            x = x.pow(2, n)
            return false if x == 1
            break if x == n - 1
            end
            return false if x != n - 1
        end
        true
    end

    private
    def generate_open_key(e, phi)
        result = e

        while (result < phi)
            puts gcdex(e, phi) == 1
            if(gcdex(e, phi) == 1)
                break;
            else
                result += 1;
            end
        end

        result
    end
    
    def generateRandomNumbers
        p,q = 0
    
        while true
            p = rand(1...100)
    
            if miller_rabin_prime?(p, K)
                break;
            end
        end
        
        iterator = 0;
    
        while true
            iterator += 1;
            q = p + iterator
    
            if miller_rabin_prime?(q, K)
                break;
            end
        end
    
        return p, q
    end
    
    def gcdex(a, b)
        if (a == 0)
            @@x = 0;
            @@y = 1;
            return b;
        end
          
        gcd = gcdex(b % a, a);
        x1 = @@x;
        y1 = @@y;
    
        c = b / a
    
        @@x = y1 - c.floor() * x1;
        @@y = x1;
      
        return gcd;
    end

    def inverse_element(a, m)
        g = gcdex(a, m);
        if (g != 1)
            puts "Inverse doesn't exist"
        else
            res = (@@x % m + m) % m;
            return res
        end
    end
end
rsa = RSA.new()

test_number = 12

puts "Тест простоти числа: #{test_number}"
puts "#{rsa.miller_rabin_prime?(test_number, K) ? "Число просте" : "Число складене"}"

puts "-------------------------"

message = 18

puts "Повідомлення: #{message}"
encrypted_text = rsa.encryption(message)

puts "Зашифроване повідомлення: #{encrypted_text}"
decrypted_text = rsa.decription(encrypted_text)

puts "Розшифроване повідомлення: #{decrypted_text}"

File: lab5.js
RELIABILITY = 45;

const millerRabinTest = (number) => {
    if(number < 3 || number % 2 === 0 ) {
        return false
    }

    number = BigInt(number);

    if( (number === 2n ) || ( number === 3n )) {
        return true;
    }

    if( (number < 2n) || ( (number % 2n) === 0 )) {
        return false;
    }

    let s = 0n, d = number - 1n;
	while (d & 1n) {
		d /= 2n;
		s++;
	}

    for(let i = 0; i < RELIABILITY; i++ ) {
        let startNewLoop = false;
        const a = BigInt(getRandomInt(Number(number) - 1));
        let x = BigInt( (a ** d) % number );

        if( (x === 1n) || (x === (number - 1n)) ) {
            continue;
        }

        for(let r = 1; r < s; r++) {
            x = (x ** 2n) % number;
            if(x === 1n) {
                return false;
            } else if ( x === (number - 1n) ) {
                startNewLoop = true;
                break;
            }
        }

        if(startNewLoop === false) {
            return false;
        }
    }
    return true;
}

const getRandomInt = number => Math.floor(Math.random() * Math.floor(number)) + 1;

const generateRandomNumbers = () => {
    let p, q = 0;

    while(true) {
        p = getRandomInt(Number(100) - 1);

        if(millerRabinTest(p)) {
            break;
        }
    }

    let iterator = 0;
    while(true) {
        iterator += 1;
        q = p + iterator

        if(millerRabinTest(q)) {
            break;
        }
    }

    return [p, q];
}

const gcdex = (a, b) => {
    let x, y;

    if (a == 0) {
        x = 0;
        y = 1;
        return b;
    }
      
    const gcd = gcdex(b % a, a);
    const x1 = x;
    const y1 = y;
 
    x = y1 - Math.floor(b / a) * x1;
    y = x1;
  
    return gcd;
}

// const [p, q] = generateRandomNumbers();
const [p, q] = [53, 17];
const n = p*q;
let e = 2;
const phi = (p-1)*(q-1);

console.log(phi);

while (e < phi) {
    if(gcdex(e, phi) === 1) {
        break;
    } else {
        e += 1;
    }
}

console.log(e);

let k = 2;
const d = (1 + (k*phi) ) / e;

const msg = 18.0;

console.log("Message data = ", msg)
 
//  Encryption c = (msg ^ e) % n
let c = Math.pow(msg, e)
c = c % n
console.log("Encrypted data = ", c)
 
//  Decryption m = (c ^ d) % n
let m = Math.pow(c, d)
m = m % n
console.log("Original Message Sent = ", m)